syntax="proto3";

package marketAccess;

// the enum definitions need to match the order in the enums of Order.h

enum orderType {
  FillOrKill = 0;
  GoodTilCancelled = 1;
}

enum orderDirection {
  buy = 0;
  sell = 1;
}

service Communication {
  rpc DisplayRequest(DisplayParameters) returns (OrderBookContent) {}

  rpc DeleteRequest(DeletionParameters) returns (Confirmation){}

  rpc InsertionRequest(InsertionParameters) returns (Confirmation){}

  rpc UpdateRequest(UpdateParameters) returns (Confirmation){}
}

message DisplayParameters{
  uint64 info = 1;
}

message OrderBookContent{
  string info = 1;
  bool validation = 2;
  optional string orderbook = 3;
  optional string comment = 4;
}

message DeletionParameters{
  uint64 info = 1;
  int32 userID = 2;
  uint64 boID = 3;
}

message InsertionParameters{
  uint64 info = 1;
  int32 userID = 2;
  double price = 3;
  double volume = 4;
  orderDirection buyOrSell = 5;
  orderType boType = 6;
}

message UpdateParameters{
  uint64 info = 1;
  int32 userID = 2;
  uint64 boID = 3;
  double price = 4;
  double volume = 5;
  orderDirection buyOrSell = 6;
  orderType boType = 7;
  int64 updatedOrderID = 8;
}

message Confirmation{
  string info = 1;
  bool validation = 2;
  optional uint64 boID = 3;
  optional string comment = 4;
}